### PR TITLE
Enhance/6261 key metrics settings toggle

### PR DESCRIPTION
**Pull request by kuasha420**
## Summary

Addresses issue:

- #6261

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: assets/js/googlesitekit/datastore/user/key-metrics.js [hunk 1]
```diff
@@ -130,64 +134,53 @@ const baseReducer = ( state, { type, payload } ) => {
...
 	 * @param {Object} state Data store's state.
-	 * @return {(Object|undefined)} Key metrics settings.
+	 * @return {(Object|undefined)} Key metrics settings. Returns `undefined` if not loaded.
 	 */
-	getUserPickedMetrics( state ) {
-		return state.keyMetrics;
+	getKeyMetricsSettings( state ) {
+		return state.keyMetricsSettings;
 	},
 };
```
[the remark this answers is not in the document]
Comment (kuasha420) [reply assets/js/googlesitekit/datastore/user/key-metrics.js:196]: Added a new selector and updated the widgets. Cheers.

File: assets/js/modules/analytics-4/components/widgets/EngagedTrafficSourceWidget.js [hunk 2]
```diff
@@ -16,6 +16,22 @@
...
+import { CORE_USER } from '../../../../googlesitekit/datastore/user/constants';
+
+const { useSelect } = Data;
+
 export default function EngagedTrafficSourceWidget() {
+	const keyMetricsWidgetHidden = useSelect( ( select ) =>
+		select( CORE_USER ).isKeyMetricsWidgetHidden()
+	);
+
+	if ( keyMetricsWidgetHidden !== false ) {
```
Comment (kuasha420) [inline assets/js/modules/analytics-4/components/widgets/EngagedTrafficSourceWidget.js:32]: Don't personally like `!==`, but this should be fine.

File: assets/js/modules/analytics-4/components/widgets/EngagedTrafficSourceWidget.test.js [hunk 3]
```diff
@@ -16,12 +16,34 @@
...
+	coreKeyMetricsEndpointRegExp,
+	setupRegistryKeyMetricsWidgetHidden,
+	setupRegistryKeyMetricsWidgetNotHidden,
+} from '../../../../util/key-metrics';
 import EngagedTrafficSourceWidget from './EngagedTrafficSourceWidget';
 
 describe( 'EngagedTrafficSourceWidget', () => {
-	it( 'should render the widget', () => {
-		const { getByText } = render( <EngagedTrafficSourceWidget /> );
+	it( 'should not render anything when isKeyMetricsWidgetHidden is not loaded', () => {
```
Comment (kuasha420) [inline assets/js/modules/analytics-4/components/widgets/EngagedTrafficSourceWidget.test.js:31]: Because all the KM widgets are barren at the moment, this tests looks repetitive and useless, however once they are populated, this may be deemed useful or removed if they don't provide any value.

File: assets/js/modules/analytics-4/components/widgets/EngagedTrafficSourceWidget.js [hunk 4]
```diff
@@ -16,6 +16,26 @@
...
-	return <div>TODO: UI for EngagedTrafficSourceWidget</div>;
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+import { CORE_USER } from '../../../../googlesitekit/datastore/user/constants';
+
+const { useSelect } = Data;
+
+export default function EngagedTrafficSourceWidget( { Widget, WidgetNull } ) {
```
Comment (eugene-manuilov) [inline assets/js/modules/analytics-4/components/widgets/EngagedTrafficSourceWidget.js:27]: Please, add prop types for this widget and for all other widgets where new props are added.

File: assets/js/util/key-metrics.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
+/**
+ * Key Metrics test utility functions.
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/util/key-metrics.js:2]: There is the `tests/js/utils.js` file for this. Please, rework your functions from this file to work as one `provideKeyMetrics` function that works similarly to other `provide*` functions that we have.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @kuasha420. This is a good start, but there are things that we need address before merging this PR. Please, take a look at my comments.

File: assets/js/components/settings/SettingsAdmin.js [hunk 6]
```diff
@@ -80,29 +81,46 @@ export default function SettingsAdmin() {
...
-							</div>
-						</Layout>
-					) }
-
-					{ isUserInputCompleted === false && (
-						<UserInputSettings isDismissible={ false } rounded />
-					) }
+								) }
+								{ isUserInputCompleted === false && (
+									<div className="googlesitekit-user-input__notification">
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/components/settings/SettingsAdmin.js:102]: Ok, I see you added a comment for this. Please, create a new issue with your comment and revert this change because we need a more holistic approach to solve that problem since that notification is shown as a banner as well and we need to make sure that we don't show that notification unnecessarily.

File: assets/js/googlesitekit/widgets/datastore/key-metrics.js
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/googlesitekit/widgets/datastore/key-metrics.js:1]: @jimmymadon, spreading the functionality from one domain between multiple data stores is not a good design choice. It makes the codebase harder to understand and maintain. Much better to keep everything that belongs to one domain in the same place.

> My idea here was that the answer to "Which widgets should be displayed?" ...

I haven't found the particular idea how exactly widgets would be rendered and controlled in the design doc (give me a link to it please). In my understanding we don't need to answer this question. Widgets rendering is a responsibility of key metric widgets themselves and they should answer on the question whether they should be rendered or not based on the currently selected key metrics.

I would say the question we need to answer is `As a key metric widget X, do I need to render my content or not?`.

File: assets/js/components/settings/SettingsAdmin.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (kuasha420) [reply assets/js/components/settings/SettingsAdmin.js:102]: Hi, the `UserInputSettings` will be removed as part of #6210, and is only displayed in Dashboard (not in Settings). If I revert this change, then there will be multiple titles which is not desired and does not satisfy the ticket. I have updated the `SettingsAdmin` to make proper usage of Grid, Row and Cells and we can further tweak this in a follow up issue. 

Reverting this and bringing back `UserInputSettings` here will have Two 'Key metrics' section which is not good.

File: assets/js/components/settings/SettingsKeyMetrics.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+		await saveKeyMetricsSettings();
+	}, [
+		keyMetricsWidgetHidden,
+		saveKeyMetricsSettings,
+		setKeyMetricsSetting,
+	] );
+
+	if ( ! keyMetricsWidgetHidden === undefined || ! keyMetrics?.length ) {
+		return null;
+	}
```
Comment (eugene-manuilov) [inline assets/js/components/settings/SettingsKeyMetrics.js:61]: Wondering whether we should return the disabled switch instead of null to avoid the cumulative layout shift if key metrics settings are loaded with a little delay. 🤔 Wdyt?

File: assets/js/components/settings/SettingsAdmin.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/components/settings/SettingsAdmin.js]: > the `UserInputSettings` will be removed as part of #6210, and is only displayed in Dashboard (not in Settings).

I checked that ticket and don't see where it says that that component should be removed. Is that the right ticket?

> If I revert this change, then there will be multiple titles which is not desired ...

Which multiple titles do you mean? This ticket is about adding a new toggle to the admin settings that controls whether to display key metrics on the dashboard page. Which multiple titles we will see in that toggle?

> ... and does not satisfy the ticket

Where in the ticket it says that we need to remove that component? I see IB says `Update assets/js/components/settings/SettingsAdmin.js to include SettingsKeyMetrics before UserInputPreview.`, but there is nothing else regarding that component.

> Reverting this and bringing back UserInputSettings here will have Two 'Key metrics' section which is not good.

I believe @jimmymadon knows about this and there is a ticket (or will be created a ticket) to resolve this. Am I right, Jimmy?

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @kuasha420. Almost looks good to me. Left two more comments.

File: assets/js/components/settings/SettingsKeyMetrics.js (hunk 7 above)
Comment (kuasha420) [reply assets/js/components/settings/SettingsKeyMetrics.js:61]: Thanks, @eugene-manuilov. 

I like the idea of there being a disabled switch to prevent layout shifts. However, it could be confusing to the user if there's no Key Metrics widget returned (ie. user have not chosen key metrics manually or answered the User Input questions) as in that case, the switch will remain disabled without any indication on why. 

I'd prefer to keep this as is and explore this idea in a follow up enhancement issue, where besides the disabled switch, possibly a tooltip or popover could be used to clearly convey the intent to the user. 

cc @jimmymadon

Comment (eugene-manuilov) [reply assets/js/components/settings/SettingsKeyMetrics.js:61]: Good point. Makes sense. Thanks.

File: assets/js/components/settings/SettingsAdmin.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (kuasha420) [reply assets/js/components/settings/SettingsAdmin.js]: > I checked that ticket and don't see where it says that that component should be removed. Is that the right ticket?

It doesn't specifically say that, yes. However, the way it will be redesigned in that issue will make the notification incompatible or overkill in its usage in `SettingsAdmin` because we will not need it in settings. See [here](https://www.figma.com/file/CdtrHB0HAjfGRWTnTMMQI0/Key-Metrics?type=design&node-id=1980-63901&t=ZHEhseYz59lKIEOb-0) for the design. 

> Where in the ticket it says that we need to remove that component? I see IB says `Update assets/js/components/settings/SettingsAdmin.js to include SettingsKeyMetrics before UserInputPreview.`, but there is nothing else regarding that component.

In the same design, you can see that the toggle should be displayed even if the user has not answered the user input questions, because user may have hand-picked the metrics. If we follow the IB and put the toggle before `UserInputPreview`, it will not be displayed when user has not answered the questions which is not correct. 

This is admittedly a bit hard to follow, but I think reverting the `SettingsAdmin` to `develop` and just adding the goggle before `UserInputPreview` is less correct than the current test, and we can definitely refine this in the future if needs be. However, if you still feel strongly in favor or a revert here, I'll do it. Just note that in that case, the toggle will not be displayed in all cases. 

Cheers.

**Comment by kuasha420** [comment]
@eugene-manuilov I've added more context on the changes. Also the VRT issue was fixed after merging develop. Please, let me know what you think. Cheers!

File: assets/js/components/settings/SettingsAdmin.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/components/settings/SettingsAdmin.js]: > However, the way it will be redesigned in that issue will make the notification incompatible or overkill in its usage in `SettingsAdmin` because we will not need it in settings.

Ok, then, could you please update that ticket to explicitly say that we need to remove `UserInputSettings` because otherwise it is just your assumption that it will be removed in that ticket and an engineer that will work on it may not know about it and will not delete that. Once that ticket IB is updated, I'll approve this ticket.

[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/components/settings/SettingsAdmin.js]: Thanks, @kuasha420.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
@kuasha420, could you please fix merge conflicts?

**Comment by hussain-t** [comment]
@kuasha420 @eugene-manuilov, I think there is confusion regarding [#6210](https://github.com/google/site-kit-wp/issues/6210). The feature description clearly says:
>The Key metrics Setup CTA placeholder component created in [#6209](https://github.com/google/site-kit-wp/issues/6209) should be designed fully in this issue.

`KeyMetricsSetupCTAWidget` was the only component added in [#6209](https://github.com/google/site-kit-wp/issues/6209). So, as part of 6210, the `KeyMetricsSetupCTAWidget` component should be designed, not `UserInputSettings`. cc: @jimmymadon

**Comment by kuasha420** [comment]
@eugene-manuilov all sorted. Please have a look. Cheers!

**Comment by hussain-t** [comment]
> @hussain-t Originally, the `UserInputSettings` component contained the "KeyMetrics Setup CTA Banner" to originally ask users to answer the questionnaire. This component was being used in two places, the Dashboard and the Admin Settings panel. This was implemented as a BannerNotification based on the requirements at that time as part of User Input V1 epic. So essentially, all we have to do is to "refactor" (simply delete) the old `UserInputSettings` component and design the `KeyMetricsSetupCTAWidget` as per the ACs of #6210.

Thanks for the clarification, @jimmymadon!

**Review by eugene-manuilov — APPROVED** [review]
Thanks, @kuasha420. LGTM!
